jupyter_client/manager.py

```diff
@@ -507,6 +514,11 @@ def _send_kernel_sigterm(self):
                     self.kernel.terminate()
                 elif hasattr(signal, "SIGTERM"):
                     self.signal_kernel(signal.SIGTERM)
+                else:
+                    self.log.debug(
+                        "Cannot set term signal to kernel, no"
```

Comment:
I see we prefer `SIGKILL` over `kill()` and assume that `kill()` exists.  Should we do the same with `SIGTERM` and `terminate()`? (I suspect this is due to Windows not having `SIGTERM`?) 

Actually, given that `Popen` is well-documented to use `SIGKILL` and `SIGTERM` for `kill()` and `terminate()`, respectively, and there don't appear to be python version compatibility issues, I think we'd be fine just calling the desired methods directly.  This would then forgo the need to log anything and remove the conditional attribute checks.

Reply:
Diggning in:

I don't think that's compatibility issue, signal_kernel signal the process group, not the kernel alone.
Sigterm can be handled by a subprocess, not sigkill.

Signal_kernel send the signal to the process group (when possible) instead of only the single process, so if the kernel has children you don't want to send sigterm to the (grand)children, put leave the child process a chance to clean up it children, in case there is some cleanup logic.

Thus `terminate()` is prefered, and if you can't you sigterm  the process group. 

Kill is the opposite, as the kernel has no chance to terminate children by catching sigkill, you want to tell them to die  as well. See #314 

I guess in a perfect world you would
 - 1 shutdown request
 - 2 wait
 - 3 term child
 - 4 wait
    -  4b if don't respond term grand children
 - 5 wait
     - 5b kill everybody. 

we just skip 4,4b, and replace 3 by "term child, if you can't term everybody".

Reply:
Good point on the process group portion of things.  I had conflated `signal_kernel()` with `send_signal()`.

Reply:
Yeah, i I had to dig deep in the code and history to figure that one :-)